Documentation/Heatington/Controllers/Serializers/CsvController.md

```diff
@@ -0,0 +1,71 @@
+
+Deserializes rawData based on the CSV format defined in RFC4180.<br>
+Decided to implement a parser over the original idea of using Regex to match for the data. The best regex foud was `/(?:,|\n|^)("(?:(?:"")*[^"]*)*"|[^",\n]*|(?:\n|$))/g` (https://gist.github.com/awwsmm/886ac0ce0cef517ad7092915f708175f), but opted agianst using it for the following reasons:
+- Even this regex is not perfect and requires some post processing, for example it includes the outer double quotes in an entry like `"a"`, so they would need to be removed. And `""` which is a way to escape a `"` in an entry inside quotes, is not possible to parse with any regex, since regex is used for pattern matching not parsing text. Therefore a lot of post processing of the data would still be required.
+- It is hard to read, understand and therefore debug.
+- Since we would only find mathces that comply with the CSV format we wouldn't know if the rawData is in the correct format in the first place and couldn't throw specific Exceptions like `throw new Exception("Entries can't start with new line.");`
+
+The CSV parser was implemented as a finite-state machine which is illistruated in the following UML diagram:
+![state machine uml diagram](sp_state_machine_UML.png)
+Side note (as it is not denoted in the diagram): Upon entering a state the next character of the string is being examined.
+
+### Serialize
+`public static string Serialize(CsvData data, bool includeHeaderIfNotNull = true)`
+
+data: the data to be serialized into CSV format<br>
+includeHeaderIfNotNull: whether to include the header in the top row of the serialized data if it exists<br>
+Returns a string with the serialized data.
+
+Serializes data based on the CSV format defined in RFC4180.<br>
+Only uses `"` characters when needed.
+
+## CsvData
+A class that holds a deserialized form of CSV data. That makes it possible to read your deserialized data as plain text or to convert it into a list of objects fit to hold that data.
+
+### Properties
+
+`public List<string[]> Table { get; private set; }`
+
+`public string[]? Header { get; private set; }`
+
+Table: holds the CSV data as plain text organized a list of arrays.<br>
+Header: holds the header as plain text in an array.
+
+### Constructor
+
+`public CsvData(List<string[]> data, string[]? header = null)`
+
+data: a table of deserialized data as plain text<br>
+header: holds entries of the CSV header
+
+Methods:
+### Create\<T\>
+
+`public static CsvData Create<T>(List<T> data, string[]? header = null)`
+
+data: a list of objects to be represented as CSV data<br>
+header: a default header to be used<br>
+Returns the CsvData created from data and header
+
+If header is null a header is created from the list of public properties of the T type. The CSV data will contain the values of the public properties of the T type turned to string via .ToString(), in case a property is null an empty string will be used as an entry.
```

Comment:
I may be wrong, but I do believe this format is better for readabilty. The same is for another methods in this documentation.
```suggestion
### Create\<T\>

`public static CsvData Create<T>(List<T> data, string[]? header = null)`

Parameters:

- `data` (List<T>): this is a list of objects that you want to represent as CSV data. 
- `header` (string[]?, optional): a default header to be used. Default is null.

Returns:

- CsvData: Returns the `CsvData` created from `data` and `header`.
```

```